From dbef8ccd10476116af944bde1224182821c9c7f0 Mon Sep 17 00:00:00 2001
From: Ravikanth Tuniki <ravikanth.tuniki@amd.com>
Date: Mon, 6 Jan 2025 22:45:06 -1200
Subject: [PATCH 096/110] net: axienet: Add EOE GRO support for RX

Generic receive offload engine receives the segmented packets from MAC
and then accumulates the data to 4KB pages. These 4KB pages are provided
to the driver via the BDs. These pages are consolidated by the driver into
64KB segments and passed to the stack. This helps achieve better throughput
in the Receive direction for high speed Ethernet interfaces.

Signed-off-by: Ravikanth Tuniki <ravikanth.tuniki@amd.com>
Signed-off-by: Suraj Gupta <suraj.gupta2@amd.com>
Reviewed-by: Radhey Shyam Pandey <radhey.shyam.pandey@amd.com>
State: pending
Message-ID: <20250107104507.359136-29-suraj.gupta2@amd.com>
---
 drivers/net/ethernet/xilinx/xilinx_axienet.h  |   2 +
 .../net/ethernet/xilinx/xilinx_axienet_eoe.c  | 171 +++++++++++++++++-
 .../net/ethernet/xilinx/xilinx_axienet_eoe.h  |  69 +++++++
 .../net/ethernet/xilinx/xilinx_axienet_main.c |  85 ++++++++-
 4 files changed, 325 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet.h b/drivers/net/ethernet/xilinx/xilinx_axienet.h
index 029dc5fb28a10..de0be07928f49 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet.h
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet.h
@@ -990,6 +990,7 @@ struct skbuf_dma_descriptor {
  * @eoe_regs: Ethernet offload IP base address.
  * @eoe_connected: Tells whether ethernet offload IP is connected to Ethernet IP.
  * @eoe_features: EOE IP supported configuration.
+ * @inetaddr_notifier: Notifier callback function for specific event.
  */
 struct axienet_local {
 	struct net_device *ndev;
@@ -1105,6 +1106,7 @@ struct axienet_local {
 	void __iomem *eoe_regs;
 	bool eoe_connected;
 	u32 eoe_features;
+	struct notifier_block inetaddr_notifier;
 };
 
 /**
diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.c b/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.c
index 417e9bbe2c766..83217548579c2 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.c
@@ -9,6 +9,10 @@
 
 #include <linux/of_address.h>
 #include <linux/platform_device.h>
+#include <linux/skbuff.h>
+#include <linux/udp.h>
+#include <linux/tcp.h>
+#include <linux/ip.h>
 
 #include "xilinx_axienet_eoe.h"
 
@@ -21,7 +25,8 @@
  *
  * This is the probe routine for Ethernet Offload Engine and called when
  * EOE is connected to Ethernet IP. It allocates the address space
- * for EOE.
+ * for EOE. Parses through device tree and updates Tx and RX offload features
+ * in netdev and axiethernet private structure respectively.
  */
 int axienet_eoe_probe(struct platform_device *pdev)
 {
@@ -44,6 +49,51 @@ int axienet_eoe_probe(struct platform_device *pdev)
 	if (IS_ERR(lp->eoe_regs))
 		return dev_err_probe(&pdev->dev, PTR_ERR(lp->eoe_regs), "couldn't map EOE regs\n");
 
+	ret = of_property_read_u32(pdev->dev.of_node, "xlnx,tx-hw-offload", &value);
+	if (!ret) {
+		dev_dbg(&pdev->dev, "xlnx,tx-hw-offload %d\n", value);
+
+		switch (value) {
+		case 0:
+			break;
+
+		case 1:
+			/* Can checksum Tx UDP over IPv4. */
+			ndev->features |= NETIF_F_IP_CSUM;
+			ndev->hw_features |= NETIF_F_IP_CSUM;
+			break;
+
+		case 2:
+			ndev->features |= NETIF_F_IP_CSUM | NETIF_F_GSO_UDP_L4;
+			ndev->hw_features |= NETIF_F_IP_CSUM | NETIF_F_GSO_UDP_L4;
+			break;
+
+		default:
+			dev_warn(&pdev->dev, "xlnx,tx-hw-offload: %d is an invalid value\n", value);
+			return -EINVAL;
+		}
+	}
+
+	ret = of_property_read_u32(pdev->dev.of_node, "xlnx,rx-hw-offload", &value);
+	if (!ret) {
+		dev_dbg(&pdev->dev, "xlnx,rx-hw-offload %d\n", value);
+
+		switch (value) {
+		case 0:
+			lp->eoe_features |= RX_HW_NO_OFFLOAD;
+			break;
+		case 1:
+			lp->eoe_features |= RX_HW_CSO;
+			break;
+		case 2:
+			lp->eoe_features |= RX_HW_UDP_GRO;
+			break;
+		default:
+			dev_warn(&pdev->dev, "xlnx,rx-hw-offload: %d is an invalid value\n", value);
+			return -EINVAL;
+		}
+	}
+
 	return 0;
 }
 
@@ -169,3 +219,122 @@ void __maybe_unused axienet_eoe_mcdma_gro_bd_free(struct net_device *ndev,
 
 	q->rxq_bd_v = NULL;
 }
+
+int axienet_eoe_recv_gro(struct net_device *ndev, int budget,
+			 struct axienet_dma_q *q)
+{
+	struct axienet_local *lp = netdev_priv(ndev);
+	static struct sk_buff *skb[XAE_MAX_QUEUES];
+	static u32 rx_data[XAE_MAX_QUEUES];
+	u32 length, packets = 0, size = 0;
+	unsigned int numbdfree = 0;
+	struct aximcdma_bd *cur_p;
+	dma_addr_t tail_p = 0;
+	struct iphdr *iphdr;
+	struct page *page;
+	struct udphdr *uh;
+	void *page_addr;
+
+	/* Get relevat BD status value */
+	rmb();
+	cur_p = &q->rxq_bd_v[q->rx_bd_ci];
+
+	while ((numbdfree < budget) &&
+	       (cur_p->status & XAXIDMA_BD_STS_COMPLETE_MASK)) {
+		tail_p = q->rx_bd_p + sizeof(*q->rxq_bd_v) * q->rx_bd_ci;
+		dma_unmap_page(ndev->dev.parent, cur_p->phys, PAGE_SIZE,
+			       DMA_FROM_DEVICE);
+
+		length = cur_p->status & XAXIDMA_BD_STS_ACTUAL_LEN_MASK;
+
+		page = (struct page *)cur_p->page;
+		if (!page) {
+			netdev_err(ndev, "Page is Not Defined\n");
+			break;
+		}
+
+		page_addr = page_address(page);
+
+		rx_data[q->chan_id - 1] += length;
+
+		if (skb[q->chan_id - 1]) {
+			skb_add_rx_frag(skb[q->chan_id - 1],
+					skb_shinfo(skb[q->chan_id - 1])->nr_frags,
+					page, 0, length, rx_data[q->chan_id - 1]);
+		}
+
+		if (((cur_p->app0 & XEOE_UDP_GRO_RXSOP_MASK) >> XEOE_UDP_GRO_RXSOP_SHIFT)) {
+			/* Allocate new skb and update in BD */
+			skb[q->chan_id - 1] = netdev_alloc_skb(ndev, length);
+			memcpy(skb[q->chan_id - 1]->data, page_addr, length);
+			skb_put(skb[q->chan_id - 1], length);
+			put_page(page);
+		} else if (((cur_p->app0 & XEOE_UDP_GRO_RXEOP_MASK) >> XEOE_UDP_GRO_RXEOP_SHIFT)) {
+			skb_set_network_header(skb[q->chan_id - 1], XEOE_MAC_HEADER_LENGTH);
+			iphdr = (struct iphdr *)skb_network_header(skb[q->chan_id - 1]);
+			skb_set_transport_header(skb[q->chan_id - 1],
+						 iphdr->ihl * 4 + XEOE_MAC_HEADER_LENGTH);
+			uh = (struct udphdr *)skb_transport_header(skb[q->chan_id - 1]);
+
+			/* App Fields are in Little Endian Byte Order */
+			iphdr->tot_len = htons(cur_p->app1 & XEOE_UDP_GRO_PKT_LEN_MASK);
+			iphdr->check = (__force __sum16)htons((cur_p->app1 &
+					XEOE_UDP_GRO_RX_CSUM_MASK) >> XEOE_UDP_GRO_RX_CSUM_SHIFT);
+			uh->len = htons((cur_p->app1 & XEOE_UDP_GRO_PKT_LEN_MASK) - iphdr->ihl * 4);
+			skb[q->chan_id - 1]->protocol = eth_type_trans(skb[q->chan_id - 1], ndev);
+			skb[q->chan_id - 1]->ip_summed = CHECKSUM_UNNECESSARY;
+			rx_data[q->chan_id - 1] = 0;
+			/* This will give SKB to n/w Stack */
+			if (skb_shinfo(skb[q->chan_id - 1])->nr_frags <= XEOE_UDP_GRO_MAX_FRAG) {
+				netif_receive_skb(skb[q->chan_id - 1]);
+				skb[q->chan_id - 1] = NULL;
+			}
+		}
+
+		size += length;
+		packets++;
+		/* Ensure that the skb is completely updated
+		 * prior to mapping the MCDMA
+		 */
+		wmb();
+		cur_p->status = 0;
+		cur_p->app0 = 0;
+		cur_p->app1 = 0;
+		page = alloc_pages(GFP_KERNEL, 0);
+		if (!page) {
+			netdev_err(ndev, "Page allocation failed\n");
+			break;
+		}
+		cur_p->page = page;
+		cur_p->phys = dma_map_page(ndev->dev.parent, page, 0,
+					   PAGE_SIZE, DMA_FROM_DEVICE);
+		if (unlikely(dma_mapping_error(ndev->dev.parent, cur_p->phys))) {
+			cur_p->phys = 0;
+			__free_pages(cur_p->page, 0);
+			netdev_err(ndev, "dma mapping failed\n");
+			break;
+		}
+
+		cur_p->cntrl = PAGE_SIZE;
+
+		if (++q->rx_bd_ci >= lp->rx_bd_num)
+			q->rx_bd_ci = 0;
+
+		/* Get relevat BD status value */
+		rmb();
+		cur_p = &q->rxq_bd_v[q->rx_bd_ci];
+		numbdfree++;
+	}
+
+	ndev->stats.rx_packets += packets;
+	ndev->stats.rx_bytes += size;
+	q->rxq_packets += packets;
+	q->rxq_bytes += size;
+
+	if (tail_p) {
+		axienet_dma_bdout(q, XMCDMA_CHAN_TAILDESC_OFFSET(q->chan_id) +
+				q->rx_offset, tail_p);
+	}
+
+	return numbdfree;
+}
diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.h b/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.h
index 84885c21e48ca..dd8a57aeab6e0 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.h
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.h
@@ -10,6 +10,8 @@
 
 #include "xilinx_axienet.h"
 
+#define XMCDMA_DFT_RX_THRESHOLD		16
+
 /* UDP Tx : GSO- Generic Segmentation offload APP0/APP1 HW offset */
 #define XMCDMA_APP0_TOTAL_PKT_LEN_MASK	GENMASK(23, 0)
 #define XMCDMA_APP0_PKT_HEAD_LEN_MASK	GENMASK(31, 24)
@@ -24,6 +26,34 @@
 #define XMCDMA_APP1_MSS_SIZE_SHIFT	16
 
 #define XEOE_UDP_NON_GRO_CHAN_ID	1
+#define XEOE_MAC_HEADER_LENGTH		0xe
+
+/* UDP Rx : GRO- Generic Receive Offload HW offset */
+#define XEOE_UDP_GRO_RX_COMMON_CR_OFFSET	0x10
+#define XEOE_UDP_CR_PROTOCOL			0x11
+#define XEOE_UDP_GRO_CR_OFFSET(chan_id)		(0x00 + ((chan_id) - 1) * 0x40)
+#define XEOE_UDP_GRO_SR_OFFSET(chan_id)		(0x04 + ((chan_id) - 1) * 0x40)
+#define XEOE_UDP_GRO_SRC_IP_OFFSET(chan_id)	(0x08 + ((chan_id) - 1) * 0x40)
+#define XEOE_UDP_GRO_DST_IP_OFFSET(chan_id)	(0x0C + ((chan_id) - 1) * 0x40)
+#define XEOE_UDP_GRO_PORT__OFFSET(chan_id)	(0x10 + ((chan_id) - 1) * 0x40)
+#define XEOE_UDP_GRO_FRAG		0x10000000
+#define XEOE_UDP_GRO_TUPLE		BIT(3)
+#define XEOE_UDP_GRO_CHKSUM		BIT(1)
+#define XEOE_UDP_GRO_BUSY_MASK		BIT(0)
+#define XEOE_UDP_GRO_4K_FRAG_SIZE	BIT(20)
+#define XEOE_UDP_GRO_ENABLE		BIT(0)
+
+#define XEOE_UDP_GRO_RXSOP_SHIFT	30 /* First GRO Packet */
+#define XEOE_UDP_GRO_RXEOP_SHIFT	29 /* Last GRO Packet */
+
+#define XEOE_UDP_GRO_RXSOP_MASK		BIT(30)
+#define XEOE_UDP_GRO_RXEOP_MASK		BIT(29)
+
+#define XEOE_UDP_GRO_MAX_FRAG		16
+
+#define XEOE_UDP_GRO_PKT_LEN_MASK	GENMASK(15, 0)
+#define XEOE_UDP_GRO_RX_CSUM_MASK	GENMASK(31, 16)
+#define XEOE_UDP_GRO_RX_CSUM_SHIFT	16
 
 /* EOE Features */
 #define RX_HW_NO_OFFLOAD		BIT(0)
@@ -42,6 +72,9 @@ int __maybe_unused axienet_eoe_mcdma_gro_q_init(struct net_device *ndev,
 						int i);
 void __maybe_unused axienet_eoe_mcdma_gro_bd_free(struct net_device *ndev,
 						  struct axienet_dma_q *q);
+int axienet_eoe_recv_gro(struct net_device *ndev,
+			 int budget,
+			 struct axienet_dma_q *q);
 #else
 static inline int axienet_eoe_probe(struct platform_device *pdev)
 {
@@ -67,6 +100,13 @@ static inline int __maybe_unused axienet_eoe_mcdma_gro_q_init(struct net_device
 static inline void __maybe_unused axienet_eoe_mcdma_gro_bd_free(struct net_device *ndev,
 								struct axienet_dma_q *q)
 { }
+
+static inline int axienet_eoe_recv_gro(struct net_device *ndev,
+				       int budget,
+				       struct axienet_dma_q *q)
+{
+	return 0;
+}
 #endif
 
 static inline bool axienet_eoe_is_channel_gro(struct axienet_local *lp,
@@ -75,4 +115,33 @@ static inline bool axienet_eoe_is_channel_gro(struct axienet_local *lp,
 	return ((lp->eoe_features & RX_HW_UDP_GRO) && q->chan_id != XEOE_UDP_NON_GRO_CHAN_ID);
 }
 
+/**
+ * axienet_eoe_ior - Memory mapped EOE register read
+ * @lp:		Pointer to axienet local structure
+ * @offset:	Address offset from the base address of EOE
+ *
+ * Return: The contents of the EOE register
+ *
+ * This function returns the contents of the corresponding register.
+ */
+static inline u32 axienet_eoe_ior(struct axienet_local *lp, off_t offset)
+{
+	return ioread32(lp->eoe_regs + offset);
+}
+
+/**
+ * axienet_eoe_iow - Memory mapped EOE register write
+ * @lp:		Pointer to axienet local structure
+ * @offset:	Address offset from the base address of EOE
+ * @value:	Value to be written into the EOE register
+ *
+ * This function writes the desired value into the corresponding EOE
+ * register.
+ */
+static inline void axienet_eoe_iow(struct axienet_local *lp, off_t offset,
+				   u32 value)
+{
+	iowrite32(value, lp->eoe_regs + offset);
+}
+
 #endif /* XILINX_AXIENET_EOE_H */
diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
index ca6fcc26d8d79..c1302c300f763 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
@@ -52,6 +52,7 @@
 #include <linux/ptp/ptp_xilinx.h>
 #include <linux/workqueue.h>
 #include <linux/gpio/consumer.h>
+#include <linux/inetdevice.h>
 
 #include "xilinx_axienet.h"
 #include "xilinx_axienet_eoe.h"
@@ -2349,7 +2350,12 @@ int xaxienet_rx_poll(struct napi_struct *napi, int quota)
 			dev_err(lp->dev, "Rx error 0x%x\n\r", status);
 			break;
 		}
-		work_done += axienet_recv(lp->ndev, quota - work_done, q);
+
+		if (axienet_eoe_is_channel_gro(lp, q))
+			work_done += axienet_eoe_recv_gro(lp->ndev, quota - work_done, q);
+		else
+			work_done += axienet_recv(lp->ndev, quota - work_done, q);
+
 		status = axienet_dma_in32(q, XMCDMA_CHAN_SR_OFFSET(q->chan_id) +
 					  q->rx_offset);
 	}
@@ -4597,6 +4603,49 @@ static const struct of_device_id axienet_of_match[] = {
 
 MODULE_DEVICE_TABLE(of, axienet_of_match);
 
+static int axienet_eoe_netdev_event(struct notifier_block *this, unsigned long event,
+				    void *ptr)
+{
+	struct axienet_local *lp = container_of(this, struct axienet_local,
+						inetaddr_notifier);
+	struct in_ifaddr *ifa = ptr;
+	struct axienet_dma_q *q;
+	int i;
+
+	struct net_device *ndev = ifa->ifa_dev->dev;
+
+	if (lp->ndev != ndev) {
+		dev_err(lp->dev, " ndev is not matched to configure GRO IP address\n");
+	} else {
+		switch (event) {
+		case NETDEV_UP:
+			dev_dbg(lp->dev, "%s:NETDEV_UP\n", __func__);
+			for_each_rx_dma_queue(lp, i) {
+				q = lp->dq[i];
+				if (axienet_eoe_is_channel_gro(lp, q))
+					axienet_eoe_iow(lp,
+							XEOE_UDP_GRO_DST_IP_OFFSET(q->chan_id),
+							ntohl(ifa->ifa_address));
+			}
+		break;
+		case NETDEV_DOWN:
+			dev_dbg(lp->dev, "%s:NETDEV_DOWN\n", __func__);
+			for_each_rx_dma_queue(lp, i) {
+				q = lp->dq[i];
+				if (axienet_eoe_is_channel_gro(lp, q))
+					axienet_eoe_iow(lp,
+							XEOE_UDP_GRO_DST_IP_OFFSET(q->chan_id),
+							0);
+			}
+		break;
+		default:
+			dev_err(lp->dev, "IPv4 Ethernet address is not set\n");
+		}
+	}
+
+	return NOTIFY_DONE;
+}
+
 /**
  * axienet_probe - Axi Ethernet probe function.
  * @pdev:	Pointer to platform device structure.
@@ -5108,6 +5157,21 @@ static int axienet_probe(struct platform_device *pdev)
 				goto cleanup_clk;
 			}
 		}
+
+		/* Set the TX coalesce count to 1. With offload enabled, there are not as
+		 * many interrupts as before and the interrupt for every 64KB segment needs
+		 * to be handled immediately to ensure better performance.
+		 */
+		if (ndev->hw_features & NETIF_F_GSO_UDP_L4)
+			lp->coalesce_count_tx = XMCDMA_DFT_TX_THRESHOLD;
+
+		/* Update the required thresholds for Rx HW UDP GRO
+		 * GRO receives 16 segmented data packets from MAC
+		 * and packet coalescing increases performance.
+		 */
+		if (lp->eoe_features & RX_HW_UDP_GRO)
+			lp->coalesce_count_rx = XMCDMA_DFT_RX_THRESHOLD;
+
 	} else {
 		struct xilinx_vdma_config cfg;
 		struct dma_chan *tx_chan;
@@ -5266,9 +5330,25 @@ static int axienet_probe(struct platform_device *pdev)
 
 	if (lp->axienet_config->mactype == XAXIENET_1G_10G_25G)
 		INIT_DELAYED_WORK(&lp->restart_work, speed_monitor_thread);
+	/* Register notifier for inet address additions/deletions.
+	 * It should be called after register_netdev to access the interface's
+	 * network configuration parameters.
+	 */
+
+	if (lp->eoe_features & RX_HW_UDP_GRO) {
+		lp->inetaddr_notifier.notifier_call = axienet_eoe_netdev_event;
+		ret = register_inetaddr_notifier(&lp->inetaddr_notifier);
+		if (ret) {
+			dev_err(lp->dev, "register_netdevice_notifier() error\n");
+			goto err_unregister_netdev;
+		}
+	}
 
 	return 0;
 
+err_unregister_netdev:
+	unregister_netdev(ndev);
+
 cleanup_phylink:
 	phylink_destroy(lp->phylink);
 
@@ -5301,6 +5381,9 @@ static void axienet_remove(struct platform_device *pdev)
 		netif_napi_del(&lp->dq[i]->napi_tx);
 	}
 
+	if (lp->eoe_features & RX_HW_UDP_GRO)
+		unregister_inetaddr_notifier(&lp->inetaddr_notifier);
+
 	unregister_netdev(ndev);
 	axienet_clk_disable(pdev);
 
-- 
2.43.0

